src/Microsoft.Security.Utilities.Core/CompiledHighPerformancePattern.cs

```diff
@@ -0,0 +1,285 @@
+                                         RegexOptions.Singleline;
+
+
+    public CompiledHighPerformancePattern(string signature,
+                                           int signaturePrefixLength,
+                                           int minMatchLength,
+                                           int maxMatchLength,
+                                           Regex scopedRegex)
+    {
+#if HIGH_PERFORMANCE_CODEGEN // This validation happens at compile time and needn't be repeated at runtime.
+        if (!scopedRegex.ToString().StartsWith("^"))
+        {
+            throw new ArgumentException($"The regular expression must be anchored to the beginning of the input: '{scopedRegex}'", nameof(scopedRegex));
+        }
+
+        if (signature.Length != 3 && signature.Length != 4)
+        {
+            throw new ArgumentException($"The signature must be 3 or 4 characters long: '{signature}'.", nameof(signature));
+        }
+#endif
+        char s0 = signature[0];
+        char s1 = signature[1];
+        char s2 = signature[2];
+        char s3 = signature.Length < 4 ? '\0' : signature[3];
+
+#if HIGH_PERFORMANCE_CODEGEN
+        if (s0 > 0x7F || s1 > 0x7F || s2 > 0x7F || s3 > 0x7F)
+        {
+            throw new ArgumentException($"The signature must consist entirely of ASCII characters: '{signature}'.", nameof(signature));
+        }
+#endif
+
+        int packedSignature = s0 | (s1 << 8) | (s2 << 16);
+        if (signature.Length == 4)
+        {
+            packedSignature |= (s3 << 24);
+        }
+
+        Signature = signature;
+        SignaturePrefixLength = signaturePrefixLength;
+        MinMatchLength = minMatchLength;
+        MaxMatchLength = maxMatchLength;
+        ScopedRegex = scopedRegex;
+        PackedSignature = packedSignature;
+    }
+
+#if HIGH_PERFORMANCE_CODEGEN
+    // TODO: Use Roslyn generator: https://github.com/microsoft/security-utilities/issues/152
+    /// <summary>
+    /// This is invoked by T4 via reflection to produce the generated portion of
+    /// this class, which enumerates all the patterns that implement <see
+    /// cref="IHighPerformanceScannableKey"/> and precomputes them into to <see
+    /// cref="CompiledHighPerformancePattern"/>.
+    ///
+    /// On .NET 8+, this input is further processed by the regex source generator
+    /// to pre-compile the regexes into C# source code.
+    /// </summary>
+    internal static string GenerateAdditionalCode()
+    {
+        var sb = new StringBuilder();
+        sb.AppendLine(
+            """
+            using System.Collections.Generic;
+            using System.Linq;
+            using System.Text.RegularExpressions;
+
+            namespace Microsoft.Security.Utilities;
+
+            partial class CompiledHighPerformancePattern
+            {
+            """);
+
+        // Since patterns can share regexes, we deduplicate them and assign an ID using this counter.
```

Comment:
>regexes [](http://example.com/codeflow?start=36&length=7)

in practice it's the scoped regexes, yes, that are duplicated? at a high level we expect uniqueness to complete regexes at the rule level.

Reply:
I think this is actually true even before "scoping", though "scoping" makes it true in more cases.

For example, consider these two:


https://github.com/microsoft/security-utilities/blob/0b88ee4ee6d368f2fd6607122fddb57144c7d1b3/src/Microsoft.Security.Utilities.Core/PreciselyClassifiedSecurityKeys/SEC101_178_AzureIotHubIdentifiableKey.cs#L16

https://github.com/microsoft/security-utilities/blob/0b88ee4ee6d368f2fd6607122fddb57144c7d1b3/src/Microsoft.Security.Utilities.Core/PreciselyClassifiedSecurityKeys/SEC101_179_AzureIotDeviceProvisioningIdentifiableKey.cs#L16

They have the same signature and hence the same regex. They differ only in how they post-process a regex match using the differing checksum seeds.